### PR TITLE
Player-Controlled Singularity Adminbus

### PR DESCRIPTION
**Pull request by ghost**
Title says it all, really. Under the "Fun" tab admins find a button that will open up a list of all singularities. At this menu any singularity can be turned into a deadchat controlled one. Currently, every message in deadchat will be processed. Every person has a separate timer - currently it is a 6 seconds, but I still may need to raise that even more since it's a goddamn singularity.

Edit: thanks for the gold kind stranger!
Edit 2: oh wow plat too? i cant believe this im so happy
Edit 3: To clarify, this works like twitch plays pokemon.

Done at the request of IFlashYou, the ebin cultmin everybody loves.

On the coding end, this adds /obj/machinery/singularity/deadchat_controlled that automatically connects itself to deadchat on creation. The admin command simply turns a singularity into a deadchat_controlled version. Deadchat is also gunky, so I cleaned it up a bit and added a /datum/deadchat_listener so that you can actually read what's happening in it. To add a listener, create a new /datum/deadchat_listener/child_version and add it to the global_deadchat_listeners. It will then trigger proc/deadchat_event whenever someone speaks with the ckey and message. Instead of having it automatically add itself to listeners you have to manually do it yourself so that future coders don't fuck up code by forgetting how it works. 

 * rscadd: Admins can now turn a singularity into a deadchat-plays-singularity bus. Type up, down, left, or right to move it.

### CODE REVIEW

**Comment by iFlashYou** [comment]
based

**Comment by SonixApache** [comment]
~~the adminbus literally already exists for this~~
after actually reading the contents i realize that this is not the case

this is still dumb spam-promoting shit

**Comment by nuklearcellphoneg** [comment]
Does it let you twitch chat vote?

**Comment by ghost** [comment]
It's adminbus, it's meant for memes when the gult takes 3 hours to do the sac

**Comment by FudgePucker** [comment]
11/10 best PR of the year. Speed merge when?

**Comment by kilozombie** [comment]
Interesting PR name, but I support the addition

image

**Comment by nuklearcellphoneg** [comment]
>this is still dumb spam-promoting shit

Should make it where the logs exclude the commands so as not to obscure log divers.

**Comment by kilozombie** [comment]
But then people will ERP using only the words to indicate direction
Think about how hot it'd be
(Follow) Jeremy Jerkins whines, "Up"
(Follow) Jeremy Jerkins complains, "Down"
(Follow) Jeremy Jerkins sulks, "Up"
(Follow) Jeremy Jerkins whines,  Down"

**Comment by nuklearcellphoneg** [comment]
>But then people will ERP using only the words to indicate direction

False equivalence. Those who ERP have no direction in life.

**Comment by ghost** [comment]
>Should make it where the logs exclude the commands so as not to obscure log divers.

I'm sure log divers will manage, the logs are already so much of a clusterfuck that you just need to crtl-f. I could add it if you REALLY want it to but I don't really see why.

**Comment by JedediahSmith** [comment]
Merge this shit NOW

**Comment by spoonylove** [comment]
What are some other fun things to deadchat control? Scp-183?

**Comment by oversane** [comment]
>scp-183

ah yes cant wait for player controlled sentient wire

**Comment by ShiftyRail** [comment]
Make a new investigate category with its own file and logging unless it is too much effort

File: code/modules/power/singularity/singularity.dm [hunk 1]
```diff
@@ -682,3 +674,92 @@
 	if(timestopped)
 		return 0
 	return forceMove(get_step(src,Dir))
+
+
+////////////This singularity is upgraded to be controlled by deadchat. God save us all.
+
+/datum/deadchat_listener/singulo_listener
+	name = "deadchat-controlled singularity listener"
+	var/obj/machinery/singularity/deadchat_controlled/parent
```
Comment (ShiftyRail) [inline code/modules/power/singularity/singularity.dm:684]: Set this to null in Destroy, please

**Review by ShiftyRail — COMMENTED** [review]
Don't forget to log the use of that command
Also, maybe make it require R_FUN I dunno

**Comment by ghost** [comment]
I'm unfamiliar with investigate categories, but I'll work on those other changes. What is it, exactly?

**Comment by ghost** [comment]
Added the logs. I just put the investigation log under I_SINGULO.

**Comment by Blithering** [comment]
while this is fun, the chat spam will be annoying to the one guy not trying to steer the sing into the SME. any chance of a different version that gives ghosts haunting the sing a ghostly DDR pad to try to control it with?

**Comment by ghost** [comment]
They can always turn off deadchat, but I think I'm able to do something to help people who don't want to partake in it. A DDR pad seems obscure, though. Any other suggestions?

**Comment by Blithering** [comment]
some kind of directional arrow HUD element, or something on the singularity itself only ghosts can see? hovering over the up arrow of it would cast a vote for moving up, etc.

**Comment by FudgePucker** [comment]
>it wont be fun for the one guy not trying to steer the singularity 

Yeah, the situations that this buttons being used EVERYBODY will be trying to steer the singularity.

**Comment by kilozombie** [comment]
Chat being clogged with singuposting is the whole point

Also this means I can start a placebo train by observing a loose goose and spam posting "UP" while knowing it's not doing anything, leading to everyone else in deadchat following suit

**Comment by Irradiation** [comment]
This is a thoroughly horrible idea brought to you by a thoroughly horrible person.

Wat cud possibyl goh rawng?

**Comment by ghost** [comment]
That is quite rude to iflashyou.

**Comment by iFlashYou** [comment]
@ShiftyRail merge

Edit: @GlassEclipse Does dedchat get a message when some badmin enables it?

**Comment by ghost** [comment]
Yeas, it does. I realized now that the jumpto link uses the function formatjumpto(src), and that might be an admin only jumpto. I'm not sure whether it is or not.

**Comment by SonixApache** [comment]
conflicts

**Comment by ghost** [comment]
Fixed conflict.

**Comment by ghost** [comment]
Fixed the fix to the conflict.

**Comment by ghost** [comment]
New update adds democracy mode and allows admins to set the delay for democracy and anarchy mode themselves. It can't be 0, but you could put a .1 delay. It also improves the movement and adds more pomf-flavored gunk code for you to enjoy.

**Pull request merged**